it/src/main/java/com/google/cloud/teleport/it/bigquery/DefaultBigQueryResourceManager.java

```diff
@@ -0,0 +1,332 @@
+        logInsertErrors(insertResponse.getInsertErrors());
+      }
+
+    } catch (Exception e) {
+      throw new BigQueryResourceManagerException("Failed to write to table.", e);
+    }
+
+    LOG.info(
+        "Successfully wrote {} records to {}.{}.",
+        successfullyWrittenRecords,
+        dataset.getDatasetId().getDataset(),
+        tableName);
+  }
+
+  @Override
+  public synchronized ImmutableList<FieldValueList> readTable(String tableName) {
+    Table table = getTableIfExists(tableName);
+
+    // List to store fetched rows
+    ImmutableList.Builder<FieldValueList> immutableListBuilder = ImmutableList.builder();
+
+    LOG.info("Reading all rows from {}.{}", dataset.getDatasetId().getDataset(), tableName);
+
+    // Read all the rows from the table given by tableId
+    try {
+      TableResult results = bigQuery.listTableData(table.getTableId());
+      for (FieldValueList row : results.getValues()) {
+        immutableListBuilder.add(row);
+      }
+    } catch (Exception e) {
+      throw new BigQueryResourceManagerException("Failed to read from table.", e);
+    }
+
+    ImmutableList<FieldValueList> tableRows = immutableListBuilder.build();
+    LOG.info(
+        "Loaded {} rows from {}.{}",
+        tableRows.size(),
+        dataset.getDatasetId().getDataset(),
+        tableName);
+
+    return tableRows;
+  }
+
+  @Override
+  public synchronized void cleanupAll() {
+    LOG.info("Attempting to cleanup manager.");
+    try {
+      if (dataset != null) {
+        Page<Table> tables = bigQuery.listTables(dataset.getDatasetId());
+        for (Table table : tables.iterateAll()) {
+          bigQuery.delete(
+              TableId.of(
+                  projectId, dataset.getDatasetId().getDataset(), table.getTableId().getTable()));
+        }
+        bigQuery.delete(dataset.getDatasetId());
+      }
+    } catch (Exception e) {
+      throw new BigQueryResourceManagerException("Failed to delete resources.", e);
+    }
+    LOG.info("Manager successfully cleaned up.");
+  }
+
+  /** Builder for {@link DefaultBigQueryResourceManager}. */
+  public static final class Builder {
+
+    private BigQuery bigQuery;
+    private final String testId;
+    private final String projectId;
+    private Credentials credentials;
+
+    private Builder(String testId, String projectId) {
+      this.testId = testId;
+      this.projectId = projectId;
+    }
+
+    public Builder setBigQuery(BigQuery bigQuery) {
+      this.bigQuery = bigQuery;
+      return this;
+    }
+
+    public Builder setCredentials(Credentials credentials) {
+      this.credentials = credentials;
+      return this;
+    }
+
+    public DefaultBigQueryResourceManager build() {
+      // create default bigQuery client if one was not set.
+      BigQuery bigQueryToUse;
+      if (bigQuery == null) {
+        BigQueryOptions.Builder bigQueryOptions =
+            BigQueryOptions.newBuilder().setProjectId(projectId);
+
+        // set credentials, if provided
```

Comment:
Will this work if credentials are not provided?

Reply:
A better question: Is there a case where the credentials will not be provided?

Reply:
I think there could be a case from an open-source contributor running the IT's, but in general, no I don't think there will be a case. However, since it is technically optional, I saw the typical builder pattern had setters for any optional value, so I followed that pattern. It's also consistent with the Pub/Sub RM builder.